### PR TITLE
Delay float32 upcast in ForCausalLMLoss after filtering ignore_index

### LINKED ISSUE #38452: Memory saving by upcasting logits for only non-ignored positions
### Feature request

In [`loss_utils.py`](https://github.com/huggingface/transformers/blob/main/src/transformers/loss/loss_utils.py), logits are upcasted for float32 for some losses. This can waste memory for cases where certain labels are `ignore_index`. This is especially true for fine tuning cases where one chooses to calculate loss only on the completion. They would keep label as -100 for prompt tokens and upcasting those logits would be unnecessary. We can instead call `logits.float()` after we have our final labels. This would be especially useful for `ForCausalLMLoss` as that seems to be the most likely use case.

### Motivation

When fine tuning a causal LM, one can choose to calculate loss only on the completion, thus setting labels for prompt tokens to be -100. Upcasting logits at those positions when calculating loss is not needed. Avoiding that can save memory. Most likely use case is `ForCausalLMLoss`.

### Your contribution

An example for `ForCausalLMLoss`:

```
def ForCausalLMLoss(
    logits,
    labels,
    vocab_size: int,
    num_items_in_batch: Optional[int] = None,
    ignore_index: int = -100,
    shift_labels: Optional[torch.Tensor] = None,
    **kwargs,
) -> torch.Tensor:
    # Don't upcast yet
    # logits = logits.float()

    if shift_labels is None:
        # Shift so that tokens < n predict n
        labels = nn.functional.pad(labels, (0, 1), value=ignore_index)
        shift_labels = labels[..., 1:].contiguous()   

    # Flatten the tokens
    logits = logits.view(-1, vocab_size)
    shift_labels = shift_labels.view(-1)

    # Upcast to float if we need to compute the loss to avoid potential precision issues
    # Now that we have our final labels, take only the useful logits and then upcast
    logits_truncated = logits[shift_labels != ignore_index]
    shift_labels_truncated = shift_labels[shift_labels != ignore_index]

    # Enable model parallelism
    shift_labels_truncated = shift_labels_truncated.to(logits.device)
    loss = fixed_cross_entropy(logits_truncated, shift_labels_truncated, num_items_in_batch, ignore_index, **kwargs)
    return loss
```

We can do something similar in `ForMaskedLMLoss` on line 83 instead of 77. `ForTokenClassification` does not take `ignore_index` as an argument but we can still do the same here because `fixed_cross_entropy` does take `ignore_index`.

Another alternative was to move the upcasting to inside `fixed_cross_entropy` but a few losses don't do that. So, that might change/break existing things.

Let me know if this change sounds good. I can submit a PR.

### PR DESCRIPTION
**Pull request by starcatmeow**
# What does this PR do?

This PR implements the optimization discussed in #38452, originally proposed by @harshit2997.  
Thanks for the original suggestion and discussion.

- Move the float32 upcast in `ForCausalLMLoss` to after filtering out `ignore_index` labels.
- Ensures only relevant logits are upcasted, reducing VRAM usage without affecting correctness.

Fixes #38452

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #38452
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- @ArthurZucker
- @Rocketknight1
- @harshit2997

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Hi @starcatmeow, I just took a look and I'm not sure we can accept this! Although I thought it was a good optimization in the issue at #38452, this is because I misread - I thought the code was already selecting masked labels, it was just doing it **after** casting to `float32` instead of before. In that case, doing the select first would save memory with no issues.

However, adding a mask select step when we didn't use one before introduces some problems - in particular, it makes the sizes of the output tensors data-dependent, which can force recompilations with less efficient dynamic shapes. I'm not sure it's worth it for the memory saving here! cc core maintainers @arthurzucker @cyrilvallez for their opinion

**Comment by Cyrilvallez** [comment]
Interesting idea, I'm not sure how often we expect to see `ignore_index` in the labels, cc @ArthurZucker do you have an idea? It could actually save a LOT of memory as vocabularies tend to be large now (200k+)

Mathematically, it's fully equivalent

**Comment by harshit2997** [comment]
Thanks @starcatmeow for picking this change up. @Cyrilvallez my main motivation behind proposing the change was to account for fine-tuning cases where one doesn't want to fine tune on prompt tokens. In addition, won't this also help with getting rid of padding tokens before the upcast which is a common scenario?
